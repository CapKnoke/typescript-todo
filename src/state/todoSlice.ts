import { createEntityAdapter, createSlice, EntityState, PayloadAction } from '@reduxjs/toolkit';
import { nanoid } from '@reduxjs/toolkit';
import { RootState } from './store';

export type Todo = {
  id: string,
  title: string,
  description: string | null,
  complete: boolean,
}

const todosAdapter = createEntityAdapter<Todo>();
const sliceSelectors = todosAdapter.getSelectors<EntityState<Todo>>(state => state);

const todoSlice = createSlice({
  name: 'todos',
  initialState: todosAdapter.getInitialState(),
  reducers: {
    hydrate: (state: EntityState<Todo>, action: PayloadAction<EntityState<Todo>>) => action.payload,
    addTodo: {
      reducer: todosAdapter.addOne,
      prepare: (formValues: { title: string, description: string }) => {
        return {
          payload: {
            id: nanoid(),
            title: formValues.title,
            description: formValues.description || null,
            complete: false,
          },
        };
      },
    },
    completeTodo(state: EntityState<Todo>, action: PayloadAction<string>) {
      const todo = sliceSelectors.selectById(state, action.payload);
      todosAdapter.updateOne(state, {
        id: action.payload,
        changes: {
          complete: !todo?.complete,
        }
      });
    },
    deleteTodo: todosAdapter.removeOne,
  },
});

export const {
  hydrate,
  addTodo,
  completeTodo,
  deleteTodo,
} = todoSlice.actions;

export const todoSelectors = todosAdapter.getSelectors((state: RootState) => state.todos);
export const selectTodoById = (id: string) => (state: RootState) => todoSelectors.selectById(state, id);

export default todoSlice.reducer;