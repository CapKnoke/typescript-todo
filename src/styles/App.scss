* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

.main-content {
  margin: 1rem;
}

.header-card {
  .title-container {
    margin-top: 1rem;
    display: flex;
    flex-direction: column;
    align-items: center;
    &__clock {
      font-family: 'Roboto Mono';
    }
    &__title {
      font-size: 30pt;
    }
  }
  .registration-form {
    display: flex;
    flex-direction: column;
    gap: 1rem;
    margin: 1rem;
  }
}

.todo-list {
  .todo {
    padding-inline: 1rem;
    margin-bottom: 1rem;
    &--completed {
      order: 1;
      background-color: rgb(154, 154, 154);
      color: rgb(73, 73, 73);
      text-decoration: line-through;
      .todo__divider {
        background-color: rgb(73, 73, 73);
      }
    }
    cursor: pointer;
    .todo-header {
      min-height: 2.5rem;
      margin-block: 1rem;
      display: flex;
      justify-content: space-between;
    }
    .todo-body {
      margin-block: 1rem;
    }
  }
}

@media screen and (min-width: 635px) {
  .main-content {
    margin-inline: auto;
    max-width: 600px;
  }
}
