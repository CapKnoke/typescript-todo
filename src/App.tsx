import React, { useMemo } from 'react';
import useMediaQuery from '@mui/material/useMediaQuery';
import CssBaseline from '@mui/material/CssBaseline';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { Divider, Stack } from '@mui/material';
import { getDesignTokens }from './theme';
import { store } from './state/store';
import TodoForm from './components/TodoForm';
import TodoList from './components/TodoList';
import './styles/App.scss';

function App() {
  store.subscribe(()=>{
    localStorage.setItem('reduxState', JSON.stringify(store.getState()))
  });
  const prefersDarkMode = useMediaQuery('(prefers-color-scheme: dark)');
  const theme = useMemo(() => (
    createTheme(getDesignTokens(prefersDarkMode ? 'dark' : 'light'))
  ),[prefersDarkMode]);
  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      <Stack
        className='main-content'
        divider={<Divider orientation="horizontal" flexItem />}
        gap={2}
      >
        <TodoForm />  
        <TodoList />
      </Stack>
    </ThemeProvider>
  );
}

export default App;
