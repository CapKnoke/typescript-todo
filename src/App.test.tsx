import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import App from './App';
import { store } from './state/store';
import { hydrate } from './state/todoSlice';

jest
  .useFakeTimers()
  .setSystemTime(new Date('2020-01-01T13:37:00'));

store.dispatch(hydrate({
  ids: ['1234', '4321'],
  entities: {
    '1234': {
      id: '1234',
      title: 'Test Todo 1',
      description: 'test description',
      complete: false,
    },
    '4321': {
      id: '4321',
      title: 'Test Todo 2',
      description: null,
      complete: true,
    },
  },
}));

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <App />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
