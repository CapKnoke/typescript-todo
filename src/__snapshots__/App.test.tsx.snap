// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="main-content css-1ucnu3w-MuiStack-root"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation6 header-card css-azc51k-MuiPaper-root"
  >
    <section
      className="title-container"
    >
      <h1
        className="title-container__title"
      >
        Add ToDo
      </h1>
      <span
        className="title-container__clock"
      >
        01/01/2020, 13:37:00
      </span>
    </section>
    <form
      action="submit"
      className="registration-form MuiBox-root css-0"
      onSubmit={[Function]}
    >
      <div
        className="MuiFormControl-root MuiTextField-root input-container__input css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="txtTodoItemToAdd"
          id="txtTodoItemToAdd-label"
        >
          Title
          <span
            aria-hidden={true}
            className="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id="txtTodoItemToAdd"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-1ftyaf0"
            >
              <span>
                Title
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root input-container__input css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor=":r1:"
          id=":r1:-label"
        >
          Description
        </label>
        <div
          className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id=":r1:"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-1ftyaf0"
            >
              <span>
                Description
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root registration-form__button css-jh47zj-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        id="btnAddTodo"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Add"
        type="submit"
      >
        Add
        <span
          className="MuiButton-endIcon MuiButton-iconSizeLarge css-z4rqyd-MuiButton-endIcon"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
      </button>
    </form>
  </div>
  <hr
    className="MuiDivider-root MuiDivider-fullWidth MuiDivider-flexItem css-yk10cl-MuiDivider-root"
  />
  <div
    className="todo-list css-nen11g-MuiStack-root"
    id="todoList"
  >
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation6 todo todo--toggle-completed css-azc51k-MuiPaper-root"
      id="1234"
      onClick={[Function]}
      onKeyDown={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      role="button"
      tabIndex={0}
    >
      <section
        className="todo-header"
      >
        <h1
          className="todo-header__title"
        >
          Test Todo 1
        </h1>
        <section
          className="todo-button-container"
        />
      </section>
      <hr
        className="MuiDivider-root MuiDivider-fullWidth MuiDivider-flexItem todo__divider css-yk10cl-MuiDivider-root"
      />
      <section
        className="todo-body"
      >
        <p>
          test description
        </p>
      </section>
    </div>
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation6 todo todo--toggle-completed todo--completed css-azc51k-MuiPaper-root"
      id="4321"
      onClick={[Function]}
      onKeyDown={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      role="button"
      tabIndex={0}
    >
      <section
        className="todo-header"
      >
        <h1
          className="todo-header__title"
        >
          Test Todo 2
        </h1>
        <section
          className="todo-button-container"
        >
          <button
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorError MuiIconButton-sizeMedium todo-header__button todo__button--remove css-1jo1f9u-MuiButtonBase-root-MuiIconButton-root"
            disabled={false}
            id="removeButton"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ClearIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </section>
      </section>
    </div>
  </div>
</div>
`;
